src/threading.c

```diff
@@ -953,44 +952,40 @@ void sleep_for(int secs, int nsecs)
 uint8_t check_heartbeats(uint8_t gc_state)
 {
     int hb = jl_atomic_exchange(&heartbeats, 0);
-    uint64_t curr_s = jl_hrtime() / 1e9;
 
     if (hb <= 0) {
-        // we didn't get a heartbeat in the last interval; should we report?
-        if (n_reported < n_loss_reports &&
-                curr_s - last_report_s >= report_interval_s) {
-            jl_task_t *ct = jl_current_task;
-            jl_ptls_t ptls = ct->ptls;
-
-            // exit GC-safe region to report then re-enter
-            jl_gc_safe_leave(ptls, gc_state);
-            jl_safe_printf("==== heartbeat loss ====\n");
-            jl_print_task_backtraces(0);
-            gc_state = jl_gc_safe_enter(ptls);
-
-            // we've reported
-            n_reported++;
-
-            // record the reporting time _after_ the report
-            last_report_s = jl_hrtime() / 1e9;
-
-            // double the reporting interval up to a maximum
-            if (report_interval_s < 60 * heartbeat_interval_s) {
-                report_interval_s *= 2;
+        // we didn't get a heartbeat
+        n_hbs_recvd = 0;
+        n_hbs_missed++;
+
+        // if we've printed task backtraces already, do nothing
+        if (!tasks_showed) {
+            // otherwise, at least show this message
+            jl_safe_printf("==== heartbeat loss (%ds) ====\n", n_hbs_missed * heartbeat_interval);
```

Comment:
You maybe/probably want to always print the heartbeat loss message, even if you've already shown the tasks, so that we'll get the maximum number before they recover?

Or maybe when they recover you want to print one more message like "heartbeats recovered; maximum loss duration = ..."?